base/cvd/cuttlefish/host/commands/cvd/instances/instance_database.cpp

```diff
@@ -58,8 +57,25 @@ Result<std::string> GenUniqueGroupName(const cvd::PersistentData& data) {
       group_names.size(), group_names.size() + 1);
 }
 
+bool InstanceFieldsMatch(const cvd::Instance& instance,
+                         const InstanceDatabase::Query& query) {
+  return (!query.instance_id || *query.instance_id == instance.id()) &&
+         (query.instance_names.empty() ||
+          Contains(query.instance_names, instance.name()));
+}
+
+bool GroupFieldsMatch(const cvd::InstanceGroup& group,
+                      const InstanceDatabase::Query& query) {
+  return (!query.home || *query.home == group.home_directory()) &&
+         (!query.group_name || *query.group_name == group.name());
+}
```

Comment:
May be worth mentioning those aspects in a comment on the functions?

Reply:
Added comments on these functions and on the Filter itself.

Ended up having to rewrite the `GroupMatches` function because it wasn't checking the `instance_names` property correctly (it was using an OR instead of an AND).